ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/RequestHandler.java

```diff
@@ -69,28 +70,39 @@ public RequestHandler(
     this.ksqlEngine = Objects.requireNonNull(ksqlEngine, "ksqlEngine");
     this.ksqlConfig = Objects.requireNonNull(ksqlConfig, "ksqlConfig");
     this.serviceContext = Objects.requireNonNull(serviceContext, "serviceContext");
-    this.distributor = Objects.requireNonNull(distributor, "distributor");
+    this.distributorFactory = Objects.requireNonNull(distributorFactory, "distributorFactory");
     this.commandQueueSync = Objects.requireNonNull(commandQueueSync, "commandQueueSync");
   }
 
+  /**
+   * Executes the validated statements.
+   *
+   * @param statements        the statements to execute
+   * @param propertyOverrides any custom overrides
+   * @param verifyOffset      the verifyOffset under which the statements were validated
+   *
+   * @return entities representing the execution result
+   */
   public KsqlEntityList execute(
       final List<ParsedStatement> statements,
-      final Map<String, Object> propertyOverrides
-  ) {
+      final Map<String, Object> propertyOverrides,
+      final long verifyOffset) {
     final Map<String, Object> scopedPropertyOverrides = new HashMap<>(propertyOverrides);
     final KsqlEntityList entities = new KsqlEntityList();
+    final DistributingExecutor distributor = distributorFactory.apply(verifyOffset);
     for (ParsedStatement parsed : statements) {
       final PreparedStatement<?> prepared = ksqlEngine.prepare(parsed);
       if (prepared.getStatement() instanceof RunScript) {
-        final KsqlEntityList result = executeRunScript(prepared, propertyOverrides);
+        final KsqlEntityList result = executeRunScript(prepared, propertyOverrides, verifyOffset);
```

Comment:
This current logic may likely also be wrong in the presence of a multi-statement request that includes a `RUN SCRIPT` command.  When the `RUN SCRIPT` executes it will get its own `DistributingExecutor`, which will increment its internal offset counter. But any statements after the script will use a different `DistributingExecutor` and hence there will be offset clash.